rfcs/20220815-v3.0-API-cleanup/README.md

```diff
@@ -0,0 +1,93 @@
+# RFC: v3.0 API cleanup
+
+## Introduction
+Major v3 oneDNN release is coming [soon](https://github.com/oneapi-src/oneDNN/milestone/16).
+This is a perfect opportunity to provide API improvements.
+To decrease maintenance pressure on the library, outdated APIs will be removed from the library which affects user integration code of oneDNN.
+oneDNN v3.0 will not be API/ABI backward compatible with oneDNN v2.x.
+Hence, user will need to modify their code to use oneDNN v3.0.
+Below one may find a comprehensive list of targeted changes:
+
+## List of changes
+Disclaimer: header files are copied in this folder are from oneDNN v2.7 to preserve lines to refer to exact spots.
+
+* Build:
+    * Build time option `DNNL_USE_RT_OBJECTS_IN_PRIMITIVE_CACHE` and all code supporting the `false` value will be removed.
+
+* C/C++ types:
+    * Eltwise algorithm types:
+        - [`soft_relu`](dnnl.hpp#L563) will be removed. [`soft_relu_v2`](dnnl.hpp#L565) will be renamed into `soft_relu`.
+            * End user must specify `alpha` value in a correspondent call.
+        - [`logsigmoid`](dnnl.hpp#L567) will be removed. It can be used as current [`soft_relu_v2`](dnnl.hpp#L565) with `alpha` equal to `-1`.
+            * End user must replace the algorithm with another one.
+        - [`bounded_relu`](dnnl.hpp#L561) will be removed. It can be used as current [`clip`](dnnl.hpp#L584) with `alpha` equal to `0` and `beta` equal to former `alpha`.
+            * End user must replace the algorithm with another one.
+        - [`hardswish`](dnnl.hpp#L592) will get an extension with `alpha` and `beta` parameters and follow [`hardsigmoid`](dnnl.hpp#L594) algorithm, since in many frameworks `hardswish` is defined as `x * hardsigmoid(x)`, but current version of `hardswish` has hard coded `alpha` and `beta` values to `1/6` and `1/2` while different frameworks may utilize other values.
+            * End user must specify `alpha` and `beta` values in a correspondent call.
+        - [`eltwise_gelu`](dnnl.hpp#576) will be removed as an alias to `eltwise_gelu_tanh`.
+            * End user must replace one type with another one.
+        - C enum type algorithm values will be modified.
+            * No impact for end user.
+    * Memory format tags:
+        - Values of certain plain tags will change to group them together.
+            * No impact for end user.
+    * Normalization [`use_scaleshift`](dnnl.hpp#L724) flag will be removed. This functionality is [supported](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210223-batch-norm-scale-only) through two different flags [`use_scale`](dnnl.hpp#L740) and [`use_shift`](dnnl.hpp#L745).
```

Comment:
Hi @kawakami-k, one more tag for you... Turned out that separate `use_scale` and `use_shift` are not supported in jit_uni_batch_normalization implementation for aarch64 cpu. To proceed with API changed, I decided to drop `use_shift` support from the implementation and keep `use_scale` support only, since extended support requires one more register and there's some magic I don't have time to figure out, sorry. Just heads up that both flags support is no longer there. Thank you.

Reply:
BTW I'm on intel i9...

Reply:
Hi @blueberry, the thing is that `SHIFT` argument doesn't participate in any bwd computations, it's derivative vanishes away. `DIFF_SHIFT` only does as a storage. That's the reason it was removed to not to mandate a memory object which may not be present on upper application level at that point.

Since `scaleshift` was a single piece, it didn't make any sense to split them at execution, but with separate arguments, there's less sense to keep both of them.

Reply:
Thank you @dzarukin. I am yet to see whether numbers in my high level code are broken by this change. Anyway, the documentation is wrong then, since it implies that dnnl_use_shift is relevant and that dnnl_use_scale | dnnl_use_shift replace dnnl_use_scaleshift.

Reply:
Thanks for letting me know, I'll take a look and fix the wrong part.